### PR TITLE
Display Objects, Camera management, Shaders Generator

### PR DESCRIPTION
**Pull request by wdanilo**
### Pull Request Description
- Implements https://github.com/luna/basegl/issues/85
- Partially implements https://github.com/luna/basegl/issues/79 (contain engine for generating GLSL code but the attribute names are still not managed automatically)

### Important Notes
- Please note that the files are not structured correctly. Sometimes objects are defined in wrong files, like `Camera` is defined in the `display_object.rs` file. I did not want to change it now, as the next task is about refactoring the whole file structure and it will be done during the next task.
- The file `lib/core/src/display/symbol/mesh.rs` is very in progress, you can skip reviewing it now as there is a lot of dirty code.
- The same applies to the `shapely` library - it is in progress and not used widely, it will be improved in the next iteration.

### Checklist
- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md), [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/prelude/src/lib.rs [hunk 1]
```diff
@@ -74,7 +87,119 @@ pub type PhantomData9<T1,T2,T3,T4,T5,T6,T7,T8,T9> = PhantomData2<PhantomData8<T1
...
+    pub without_phantom: T,
+    phantom: PhantomData<P>
+}
+
+impl<T, P> WithPhantom<T, P> {
+    pub fn new(without_phantom: T) -> Self {
+        let phantom = PhantomData;
+        Self { without_phantom, phantom }
+    }
+}
```
Comment (farmaazon) [inline lib/prelude/src/lib.rs]: Do we _really_ need this? why not just add phantom field to a structure?

File: lib/prelude/src/lib.rs [hunk 2]
```diff
@@ -74,7 +87,119 @@ pub type PhantomData9<T1,T2,T3,T4,T5,T6,T7,T8,T9> = PhantomData2<PhantomData8<T1
...
+
+
+
+// ==========================
+// === RefCell Management ===
+// ==========================
 
+// === RefGuard ===
+
+pub struct RefGuard<'t,Base,Data> {
```
Comment (farmaazon) [inline lib/prelude/src/lib.rs]: So we can use this in that "SharedDirtyIterator", right?

File: lib/core/src/display/scene.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+use crate::prelude::*;
+
+use crate::display::symbol::display_object::{Camera2D, DisplayObjectData};
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/scene.rs]: Nope, generated by IntelliJ. Thanks for catching.

File: lib/core/src/display/symbol/display_object.rs [hunk 4]
```diff
@@ -0,0 +1,1180 @@
...
+
+impl Camera2D {
+    pub fn mod_position<F:FnOnce(&mut Vector3<f32>)>(&self, f:F) {
+        self.rc.borrow_mut().mod_position(f)
+    }
+
+    pub fn set_position(&self, value:Vector3<f32>) {
+        self.rc.borrow_mut().set_position(value)
+    }
+}
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/symbol/display_object.rs]: What?

File: lib/core/src/tp/mod.rs [hunk 5]
```diff
@@ -0,0 +1 @@
+pub mod debug;
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/tp/mod.rs]: You lazy bastard.

File: lib/prelude/src/lib.rs (hunk 1 above)
Comment (wdanilo) [reply lib/prelude/src/lib.rs]: This allows just adding phantom field on top of any structure in a nice algebraic fashion. I'm not sure it is so usable as I thought it was. It would be definitely usable in Haskell, here I need more use cases to judge it correctly.

File: lib/code-builder/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+/// strings (including slices) and also any objects that implement the `Printer` trait.
+pub trait AddToBuilder<T> {
+    fn add_to_builder(&mut self, t:T);
+}
+
+impl<T: HasCodeRepr> AddToBuilder<&T> for CodeBuilder {
+    default fn add_to_builder(&mut self, t:&T) {
+        t.build(self)
+    }
+}
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/code-builder/src/lib.rs]: Can we? I don't know a way to do it.

File: lib/core/src/display/workspace.rs [hunk 7]
```diff
@@ -19,145 +32,224 @@ pub enum Error {
...
 }
+
+
+// ========================
+// === WorkspaceBuilder ===
+// ========================
+
+pub struct WorkspaceBuilder {
+    pub name: String 
+}
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/workspace.rs]: Probably so, yes.

File: lib/prelude/src/lib.rs (hunk 2 above)
Comment (wdanilo) [reply lib/prelude/src/lib.rs]: Yes, it seems like this is a safe generalization of that pattern. I haven't still used it much so I cannot guarantee it works in all cases.

File: lib/core/src/dirty.rs [hunk 8]
```diff
@@ -1,110 +1,539 @@
...
+where &'t T: IntoIterator {
+    type Item = <&'t T as IntoIterator>::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/dirty.rs]: I dont know why I implemented that as well. Removing.

File: lib/core/src/dirty.rs [hunk 9]
```diff
@@ -1,110 +1,539 @@
...
+    pub fn iter(&self) -> SharedDirtyFlagIter<T, OnSet> {
+        let _borrow   = self.rc.borrow();
+        let reference = unsafe { drop_lifetime(&_borrow) }; // [1]
+        let iter      = reference.data.into_iter();
+        SharedDirtyFlagIter { iter, _borrow }
+    }
+}
+
+/// Iterator guard for SharedDirtyFlag. It exposes the iterator of original
+/// structure behind the shared reference.
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/dirty.rs]: Which comment?

File: lib/core/src/dirty.rs [hunk 10]
```diff
@@ -1,110 +1,539 @@
...
+// [1] Please refer to `Prelude::drop_lifetime` docs to learn why it is safe to
+// use it here.
+impl<T, OnSet> SharedDirtyFlag<T,OnSet>
+where for<'t> &'t T: IntoIterator {
+    pub fn iter(&self) -> SharedDirtyFlagIter<T, OnSet> {
+        let _borrow   = self.rc.borrow();
+        let reference = unsafe { drop_lifetime(&_borrow) }; // [1]
+        let iter      = reference.data.into_iter();
+        SharedDirtyFlagIter { iter, _borrow }
+    }
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/dirty.rs]: Very good point. I was thinking about it already several times. There are few possible solutions here:
1. This one.
2. Exposing the ref to inner data using some getter with a name telling this is a borrow.
3. Adding a `borrow` method to dirty flags and requiring its usage on call side.
4. Cloning the data in accessor (of course bad).

I suspect you tend to option 3?

File: lib/core/src/display/symbol/buffer.rs [hunk 11]
```diff
@@ -0,0 +1,528 @@
...
+
+// === Callbacks ===
+
+pub type SetDirty    <Callback> = dirty::SharedRange<usize,Callback>;
+pub type ResizeDirty <Callback> = dirty::SharedBool<Callback>;
+
+closure! {
+fn buffer_on_resize<C:Callback0> (dirty:ResizeDirty<C>) ->
+    DataOnResize { || dirty.set() }
+}
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/symbol/buffer.rs]: I fixed this and 7 other similar places!

File: lib/core/src/display/world.rs [hunk 12]
```diff
@@ -1,143 +1,216 @@
...
 impl World {
-    pub fn new() -> Self {
-        default()
+    /// Create and initialize new world instance.
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new() -> WorldRef {
+        let world_ref  = WorldRef::new(Self::new_uninitialized());
+        let world_ref2 = world_ref.clone_rc();
+        let world_ref3 = world_ref.clone_rc();
+        with(world_ref.borrow_mut(), |mut data| {
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/world.rs]: It is usable! Without this function, it doesn't compile. Arguably, I find such nested blocks much uglier. Hmm, what to do with this one then?

File: lib/core/src/display/world.rs [hunk 13]
```diff
@@ -1,143 +1,216 @@
...
+                self.workspace_dirty.unset_all();
+                self.workspaces.iter_mut().for_each(|t| t.update());
+//            });
+//        }
+    }
+    /// Dispose the world object, cancel all handlers and events.
+    pub fn dispose(&mut self) {
+        self.self_reference = None;
+        self.update_handle  = None;
+    }
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/world.rs]: I will add your comment to the code and when we will have much more advanced world implementation and see all usage patterns we will think how to improve it.

File: lib/core/src/lib.rs [hunk 14]
```diff
@@ -1,15 +1,25 @@
...
+#![feature(set_stdio)]
 //#![warn(missing_docs)]
 
-// Lints. To be refactored after this gets resolved: https://github.com/rust-lang/cargo/issues/5034
+// Lints. To be refactored after this gets resolved:
+// https://github.com/rust-lang/cargo/issues/5034
 #![allow(clippy::option_map_unit_fn)]
+//#![feature(trace_macros)]
+//#![recursion_limit="256"]
+//trace_macros!(true);
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/lib.rs]: This is commented and kept here just to remember how to debug macros :P I know this is ugly, but this is a main file which evolves frequently now. Hmm, maybe I should add comment to tell this is for debugging purposes only.

**Comment by wdanilo** [comment]
As Danilo has a lot of work now, I would merge it now. However in case of any comments from your side, Danilo, I'd be happy to apply them!

File: lib/core/src/dirty.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (farmaazon) [reply lib/core/src/dirty.rs]: This one regarding field order

File: lib/prelude/src/lib.rs (hunk 2 above)
Comment (farmaazon) [reply lib/prelude/src/lib.rs]: So let write unit tests for it!

File: .rustfmt.toml [hunk 15]
```diff
@@ -13,7 +13,7 @@ where_single_line           = true
 condense_wildcard_suffixes  = true # `let (a,b,_,_) = x` -> `let (a,b,..) = x`
 
 # Max Width
-max_width              = 100
+max_width              = 80
```
Comment (farmaazon) [inline .rustfmt.toml]: This was intentional? Are we returning to the 80 characters per line?

File: lib/code-builder/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (farmaazon) [reply lib/code-builder/src/lib.rs]: It'll look like this way:
In Builder impl:
```rust
    pub fn add<T>(&mut self, t:&T) -> &mut Self
        where T: HasCodeRepr<T> {
        t.build(self);
        self
    }
```
And implement HasCodeRepr for String:
```rust
impl HasCodeRepr for String {
    fn build(&self, builder:&mut CodeBuilder) {
        builder.add_str(self);
    }
}
```
The only downside is that `add` function must take a reference. But this does not change much, because eventually from all Strings we're getting as_str/as_ref

File: lib/core/src/display/symbol/material/shader/builder.rs [hunk 16]
```diff
@@ -199,3 +213,22 @@ fn mk_out_name      <S:Str> (s:S) -> String { format!("out_{}", s.as_ref()) }
...
+//        typ: glsl::Type {
+//            prim: glsl::PrimType::Float,
+//            array: None
+//        }
+//    });
+//    sb.compute(&cfg,"--1--","--2--");
+//    let s  = sb.get();
+//    println!("{}",s.vertex);
+//    println!("{}",s.fragment);
+//}
```
Comment (farmaazon) [inline lib/core/src/display/symbol/material/shader/builder.rs]: I believe that we don't want to keep such commented functions in our code.

File: lib/core/src/dirty.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (farmaazon) [reply lib/core/src/dirty.rs]: Yes, a 3. looks slightly better, because we see in the usage place that something may happen.

But still I'm afraid of some complex situations, where we're iterating over flags calling some function on each iteration, which calls another functions, which calls another, which fires some callback etc. and there we dereference another copy of SharedDirtyFlag and KABOOM! 

Actually, this kind of issues is not _hard to track down_, because the stacktrace should tell everything. But we must ensure, that each code using that is covers with automatic tests, covering each fantastic corner case.

_Or_

I'm not sure, but I think that usually when we iterate over flag content we usually also clear this flag. So maybe not _cloning_ but _taking_ shared data is an option?

File: lib/code-builder/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/code-builder/src/lib.rs]: Oh, yep, it will work this way. If we have problem with the reference, then we can change `HasCodeRepr`'s `build` to one which takes ownership of self.

File: lib/core/src/dirty.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/dirty.rs]: There are two things here:

1. We are using this functionality currently in 1 place in code.
2. The idea with taking ownership is really interesting one! I'll check it out.

[the remark this answers is not in the document]
Comment (farmaazon) [reply lib/core/src/dirty.rs]: So, to explain this: It would be analogous to `drop` method of various containers in rust standard library. The dirty flags implementing `Iterator` could have `drop_all` which in essence calls `std::mem::take` on owned container and returns it as iterable. Then `SharedDirtyFlag` just calls this drop on borrow_mut without need of dropping lifetimes.
